### ISSUE TITLE
Large bedgraph file not working w/ accompanying index

### ISSUE THREAD

**Issue by blajoie** [igvteam/igv#857]
small bedgraph w/ index, chr2.  WORKS
https://1000genomes-dragen.s3.amazonaws.com/agg/test/2.bedGraph.gz
https://1000genomes-dragen.s3.amazonaws.com/agg/test/2.bedGraph.gz.tbi
![chr2-points](https://user-images.githubusercontent.com/8889420/97066174-236c1500-1568-11eb-804d-3c1426439d56.png)

medium bedgraph w/ index, chr2,3,4,5 BROKEN
https://1000genomes-dragen.s3.amazonaws.com/agg/test/2345.bedGraph.gz
https://1000genomes-dragen.s3.amazonaws.com/agg/test/2345.bedGraph.gz.tbi

large bedgraph w/ index, all chrs, BROKEN
https://1000genomes-dragen.s3.amazonaws.com/agg/test/1000genomes-3.5.7b-hg38.10.depth.samples.depth.bedGraph.gz
https://1000genomes-dragen.s3.amazonaws.com/agg/test/1000genomes-3.5.7b-hg38.10.depth.samples.depth.bedGraph.gz.tbi

![all-points](https://user-images.githubusercontent.com/8889420/97066268-e6545280-1568-11eb-9444-805ab3c7c9f0.png)

Note how with the larger files (track #2 & #3) only some data comes through (beginning of file).  **Is there a bug in the indexing code, where perhaps only 32 bit tree pointers are being used?**

Locally, tabix indexing works for any given file-size.

Ultimate goal is to use a compressed bedgraph ~22GB in size. (can share).
These tracks represent depth data from 2504 samples, binned into ~1kb bins genome wide.  Each 1kb bin would have 2504 corresponding depth values according to each samples observed depth.  Goal is to see the depth variation across all samples in a given genomic interval. 
![eee9802d-621d-4ef7-a5e3-8b1e7b7b03ba](https://user-images.githubusercontent.com/8889420/97066363-bc4f6000-1569-11eb-966c-2a6e838fe423.png)

Yes this is a big file, and yes normally bigwig is preferred, however this data represents population depth values.  We do not want the auto-binning/mean employed by big-wig, we want to see all depth values for a given bin (up to size NNN).  This file will scale with sample-count, so perhaps even bedgraph is non-ideal and we should instead explore a pre-binned (vertical histogram) type track formatted into SEG/other-format??

Cheers

**Comment by jrobinso** [comment]
I don't understand what you are trying to achieve here, but for a whole
genome view the entire file has to be read into memory whether tabix
indexed or not.    Actually it's pointless to index a file you want to view
at whole genome view.    You are probably just running out of browser
memory.    You have to set a reasonable visibility window so that a
reasonable amount of data is loaded,  you simply can't read a 22GB file to
create a whole genome view.   If you have turned off the visibility window
you might as well remove the index as well.

Perhaps I'm not understanding what you are trying to do,  could you
elaborate?

**Comment by blajoie** [comment]
Yes sorry, poor choice of the example using the small bedgraph and displaying it genome-wide. 

The goal is to re-create the view in the 2nd screenshot (top points plot), using a large (22gb) genome-wide bedgraph (which represents observed sequencing depth, genome-wide, across 2504 wgs samples.  This will be used to visually inspect sample/depth (cnv) variation across the genome.  We would allow users to view the data up to a ~3MB visibility window.

(we will tweak the visibility windows as need to fit into available memory).

The issue was that even for the medium bedgraph example or any larger sized file, data is being 'dropped' in local igv.  

Top screenshot:
track #1 has data for only chromosome 2 - all data is correctly being shown.
track #2 has data for chromosomes 2,3,4,5 - data is only shown for the first half of chr2?
track #3 has data for all chromosomes - data is only shown for the first half of chr1?

If we are encountering some sort of a memory limitation, track #2 has the same data that track #1 has for all chromosome 2 - we would expect the plotting to be cut-off further along in the track.  Also to note this test file is <90MB in size.

Our hunch, is that perhaps something is off with the reading/loading of the tabix index?  32bit/64bit tree pointers? Indexing+slicing works 100% on linux.  We are only noticing this within IGV.

Hope that helps some.

**Comment by jrobinso** [comment]
I will look into it soon.  As a rough rule fo thumb you shouldn't expect to load more than about 1mb of data max if you want good performance, 10mb max but it will be sluggish at that size.    So for a 22 GB file, assuming uniform distribution,  you shouldn't expect to load a genomic range any larger than 10kb,  if I did the math right.

I'm not sure what you are referring to wrt 32 bit / 64 bit tree pointers.   Is this some Tabix option?   Tabix uses virtual pointers use 64 bit integers.   Could you elaborate on this?

**Comment by jrobinso** [comment]
And, again,  the index isn't even used for whole genome views.    The entire file is read,  or rather an attempt is made to read the entire file.    Could you create a test case for a single chromosome?    The whole genome views are confusing the issue.   If you want to test whole genome view just comment out the indexURL.

**Comment by jrobinso** [comment]
Apologies @blajoie I read this as an igv.js issue, although clearly its posted in the IGV forum.    Tabix indexed bedgraph files are not supported in IGV desktop, they never have been. There would be some issues to overcome before they could be.   In any event they are not.  Furthermore, there is a known bug in Java WRT block gzipped files,  the Java gunzipper is only going to unzip the first block, then it detects what it thinks is the end-of-file.     For tabix block gzipped files (and also bam) a custom gunzipper, which now lives in the htsjdk, is used.   However that's not relevant as tabix bedgraph is not supported.   So what's happening here is the index is ignored,  and you are hitting the Java bgzip bug.     

To see this gunzip the data files and then recompress them with standard gzip (not bgzip).    Forget the index its not even used.


In short you can't do this with IGV desktop.    This was added at your request to igv.js.    However even there you can't support whole genome views for very large files.    As a rule of thumb calculate the genomic region in your file that would result in about 10 megabytes of compressed data loaded,  that would give you a rough guess for the maximum visibility window.    If I did the math correctly, for a 22 GB file this would correspond to about 10 kilo-bases.

**Comment by ohofmann** [comment]
I feel that's where Aaron Quinlan's (@arq5x) d4 format (https://github.com/38/d4-format, https://github.com/38/pyd4, manuscript coming soon) might be a better fit. It's something I've been looking and whether we should support it with htsget (cc @brainstorm @mlin).

**Comment by jrobinso** [comment]
@ohofmann @arq5x  I don't see how that solves the problem,  you will still need to bin the data in some fashion irrespective of how it is stored and accessed.   You simply can't load the amount of data @blajoie is anticipating.   

But as an aside supporting that via htsget would solve another problem,  I've started a discussion with Aaron and Hao on how to support this in IGV.   There are no Java readers, and distributing Rust binaries is not something we can support.

**Comment by ohofmann** [comment]
Good point. I was thinking the d4 data structure might be better suited / have a smaller memory footprint but I can see how this would not help here.

**Comment by blajoie** [comment]
> Apologies @blajoie I read this as an igv.js issue, although clearly its posted in the IGV forum. Tabix indexed bedgraph files are not supported in IGV desktop, they never have been. There would be some issues to overcome before they could be. In any event they are not. Furthermore, there is a known bug in Java WRT block gzipped files, the Java gunzipper is only going to unzip the first block, then it detects what it thinks is the end-of-file. For tabix block gzipped files (and also bam) a custom gunzipper, which now lives in the htsjdk, is used. However that's not relevant as tabix bedgraph is not supported. So what's happening here is the index is ignored, and you are hitting the Java bgzip bug.
> 
> To see this gunzip the data files and then recompress them with standard gzip (not bgzip). Forget the index its not even used.
> 
> In short you can't do this with IGV desktop. This was added at your request to igv.js. However even there you can't support whole genome views for very large files. As a rule of thumb calculate the genomic region in your file that would result in about 10 megabytes of compressed data loaded, that would give you a rough guess for the maximum visibility window. If I did the math correctly, for a 22 GB file this would correspond to about 10 kilo-bases.

Ah - yes.  Thank you @jrobinso !  That makes perfect sense with what we have been seeing in IGV (java) desktop.  Sorry for the igv vs igv.js confusion :)  And yes this does 'work' in igv.js - though as you and @ohofmann have mentioned, bedgraph as is will not scale well.

**Comment by blajoie** [comment]
We (@theandygross) have also explored commandeering the SEG format, to encode a vertical-depth histogram across all horizontal bins. This helps to show deviations from normal (diploid) depth, while maintaining some resemblance of population depth distributions at each bin.  However, we then are forced to bin the data into a single & larger bin-size to support genome-wide / whole chromosome views.  Or juggle different bin-sizes/tracks for different use-cases.

![f01f26b9-ced0-409f-ac3e-bcad1f805af3](https://user-images.githubusercontent.com/8889420/97107831-d468d180-1686-11eb-8faa-b9eda49c19f3.png)

```
0.5X	1	18250000	18750000	2
0.6X	1	18250000	18750000	2
0.7X	1	18250000	18750000	2
...
1.7X	1	18250000	18750000	2.593220339
1.8X	1	18250000	18750000	3.169491525
1.9X	1	18250000	18750000	3.406779661
2.0X	1	18250000	18750000	3.60380841
2.1X	1	18250000	18750000	3.016949153
...
3.4X	1	18250000	18750000	2
3.5X	1	18250000	18750000	2
```

An alternative we have discussed (similar to how things work in ucsc genome browser), was the ability to 'stack' multiple big-wig files into a single track.  If this were possible,  We could then produce say 30 BW files, each representing a different vertical depth bin (0.5-0.6, 0.6-0.7 ... 3.4-3.5), plot them ordered, in a single track space, and be able to view a dynamically-binned  representation of the depth data (similar to the above SEG example, one BW file for each horizontal slice through the produced 'heatmap').  For this, the auto-mean-binning of a BW format would be OK since the data is already pre-categorized into the 'heatmap' form.  

(example of depth as bedgraph, depth as bw, events)
![0d6be59d-4e34-4fe4-b929-4014cf450cad](https://user-images.githubusercontent.com/8889420/97108101-7937de80-1688-11eb-992a-9f59578c19ba.png)

Open to other ideas.  We'd love to support large data (such as population depth) in igv/igv.js!

Now that IGV-desktop BG usage is clear, would it make sense to pull this above user-scenario out as a new FR?

Cheers

**Comment by jrobinso** [comment]
@blajoie I will start another round of work on IGV desktop again this week.     You can already overlay multiple bigwig tracks (i.e. "stack"),  in both IGV desktop and igv.js.   I'm not following quite how that solves your problem but it is doable.  In your last sentence above I don't understand what "FR" is.

**Comment by blajoie** [comment]
Great!  Sorry FR means feature-request (could open a new ticket documenting the use-case/request instead of this ticket which focused on the Java bgzip bug).    Can you share an example as to how one could stack multiple BWs in a single track in igv.js?  We can explore a bit.  We had the thought awhile back - and though it could meet the use case! Certainly worth taking a second look now.  Cheers!

**Comment by jrobinso** [comment]
Example follows (for igv.js).  

```
                    {
                        name: "Merged",
                        height: 50,
                        type: "merged",
                        tracks: [
                            {
                                "type": "wig",
                                "format": "bigwig",
                                "url": "https://www.encodeproject.org/files/ENCFF000ASJ/@@download/ENCFF000ASJ.bigWig",
                                "color": "red"
                            },
                            {
                                "type": "wig",
                                "format": "bigwig",
                                "url": "https://www.encodeproject.org/files/ENCFF351WPV/@@download/ENCFF351WPV.bigWig",
                                "color": "green"
                            }
                        ]
                    }
```

**Comment by blajoie** [comment]
This is fantastic @jrobinso - many many thanks!
